storage/gc_queue_test.go

```diff
@@ -314,6 +314,138 @@ func TestGCQueueProcess(t *testing.T) {
+	defer leaktest.AfterTest(t)
+
+	const now time.Duration = 3 * 24 * time.Hour
+	const dAbandon = -2 * DefaultHeartbeatInterval
+	type spec struct {
+		status      roachpb.TransactionStatus
+		ts          time.Duration
+		heartbeatTS time.Duration
+		intentSpans []roachpb.Span
+		seqGC       bool // expect sequence cache entries removed?
+	}
+	// Describes the state of the Txn table before the test.
+	before := map[string]spec{
+		// Too young, should not touch.
+		"a": {roachpb.PENDING, now - txnCleanupThreshold + 1, 0, []roachpb.Span{{Key: roachpb.Key("q")}}, false},
+		// Old, but still heartbeat. No GC.
+		"b": {roachpb.PENDING, 0, now - txnCleanupThreshold + 1, nil, false},
+		// Old and aborted, should delete.
+		"c": {roachpb.ABORTED, now - txnCleanupThreshold - 1, 0, nil, true},
+		// Abandoned and pending, so should push and abort it successfully.
+		// But it's not old enough to actively go after the sequence cache.
+		// There's some room for optimization here by ways of sharing the
+		// transaction grooming results with the corresponding operation for
+		// the sequence cache, but that's only worthwhile in weird cases.
+		"d": {roachpb.PENDING, now - dAbandon, 0, nil, false},
+		// Committed and fresh, so no action.
+		"e": {roachpb.COMMITTED, now - txnCleanupThreshold + 1, 0, nil, false},
+		// Committed, old and intentless. Bye bye.
+		"f": {roachpb.COMMITTED, now - txnCleanupThreshold - 1, 0, nil, true},
+		// Committed and old, but with intent. The intent is resolvable,
+		// so the txn entry should be GC'ed.
+		"g": {roachpb.COMMITTED, now - txnCleanupThreshold - 1, 0,
+			[]roachpb.Span{{Key: roachpb.Key("z")}}, true},
+		// Same as the previous one, but we've rigged things so that the intent
+		// resolution here will fail and consequently no GC is expected.
+		"h": {roachpb.COMMITTED, now - txnCleanupThreshold - 1, 0,
+			[]roachpb.Span{{Key: roachpb.Key("z")}}, true},
+	}
+
+	after := map[string]*spec{}
+	for k := range before {
+		sCopy := before[k]
+		after[k] = &sCopy
+	}
+
+	// Test outcome follows, described as changes to the previous state.
+	// A status of -1 corresponds to a GC'ed record.
+	after["a"].intentSpans = nil // expect no attempts to resolve the intent
+	after["c"].status = -1
+	after["d"].status = roachpb.ABORTED
+	after["f"].status = -1
+	after["g"].status = -1
+
+	resolved := map[string][]roachpb.Span{}
+	TestingCommandFilter = func(req roachpb.Request, _ roachpb.Header) error {
+		if resArgs, ok := req.(*roachpb.ResolveIntentRequest); ok {
+			id := string(resArgs.IntentTxn.Key)
+			resolved[id] = append(resolved[id], roachpb.Span{
+				Key:    resArgs.Key,
+				EndKey: resArgs.EndKey,
+			})
+			// We've special cased one test case. Note that the intent is still
+			// counted in `resolved`.
+			if id == "h" {
+				return util.Errorf("boom")
+			}
+		}
+		return nil
+	}
+	tc := testContext{}
+	tc.Start(t)
+	defer tc.Stop()
+	defer func() { TestingCommandFilter = nil }()
```

Comment:
Move this up a few lines. `tc.Start` could panic and leave `TestingCommandFilter` in a bad state.